File: src/services/licensing.js
// Camada de serviço do Módulo de Licenciamento (Etapa 5: HWID + RSA Offline).
//
// Dentro do app Tauri, o HWID e a validação da Chave de Ativação (assinatura
// RSA sobre o HWID) são resolvidos pelos comandos Rust em
// `src-tauri/src/licensing.rs` — a chave privada nunca sai do computador do
// suporte, e o app só guarda/valida a chave pública, 100% offline.
//
// Quando a UI roda apenas no navegador (`npm run dev`, sem o shell Tauri —
// usado durante o desenvolvimento da interface), não há HWID de Windows nem
// comando nativo disponível: o HWID vira um identificador persistido em
// `localStorage` e a assinatura RSA é conferida com a mesma chave pública via
// Web Crypto (`SubtleCrypto`), para que a tela de ativação funcione de forma
// equivalente dentro e fora do app empacotado (mesmo espelhamento já usado em
// src/services/db.js).
import { isTauriRuntime, licencaLocalRepo } from './db'

const LS_HWID_SEED_KEY = 'pdv-express-db:_hwid_seed'

// Chave Serial Padrão (v1.0.5) — mesmo valor de `MASTER_SERIAL_KEY` em
// src-tauri/src/licensing.rs. Produto vendido com uma única chave de
// ativação, igual para todos os compradores, sem geração individual por
// HWID/suporte manual.
const MASTER_SERIAL_KEY = 'NSPDV-2026-PRO-BR99'

export const INVALID_KEY_MESSAGE = 'Chave inválida. Verifique o código recebido na confirmação do seu pedido.'

/** Maiúsculas, sem espaços/hífens/qualquer separador — ver `normalize_key` em src-tauri/src/licensing.rs. */
function normalizeKey(value) {
  return (value ?? '').replace(/[^a-zA-Z0-9]/g, '').toUpperCase()
}

function isMasterKey(serial) {
  return normalizeKey(serial) === normalizeKey(MASTER_SERIAL_KEY)
}

// Mesma chave pública RSA-2048 embutida no binário Rust
// (src-tauri/licensing/public_key.pem) — formato SPKI/DER em base64, pronta
// para `SubtleCrypto.importKey('spki', ...)`. Usada só no fallback de
// navegador; dentro do Tauri a verificação roda em Rust.
const PUBLIC_KEY_SPKI_BASE64 =
  'MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAhRonWvg5rdJT1/e7ZxQW' +
  'hPckPDsMUFyR1Kly618KpmUqVnWvm3fkNVQ1aqH5p9wvPLP2uKQm86JOAw0M3Oyo' +
  'pDZ7aX+6PExEcetgLXpsR8mrMu4K4do+yd3w9ODO/IxLGlNaarlrkdmx0lj4myi7' +
  'dXLxq6C0pBbMA4pez3G+yu6TBxyjrvMwoxs73sJ165kGApRIzDGzV8kwb3vDx1WG' +
  '7iRd2fqgV1w7S6jqJLuHoBAkE5EEHHUKC8w8WwBF1tlqwJkqIeMaBp6pc4S/GNqI' +
  'qt2A9XqCb3Unj+FSYIL9HyUDIY5W/GKkqcR7W2p1BVZWL4xpLSzqu9RYj7kqo9dS' +
  '0wIDAQAB'

function base64ToBytes(base64) {
  const binary = window.atob(base64)
  const bytes = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i += 1) bytes[i] = binary.charCodeAt(i)
  return bytes
}

async function sha256Hex(text) {
  const data = new TextEncoder().encode(text)
  const digest = await window.crypto.subtle.digest('SHA-256', data)
  return [...new Uint8Array(digest)].map((byte) => byte.toString(16).padStart(2, '0')).join('')
}

function readOrCreateHwidSeed() {
  try {
    const existing = window.localStorage.getItem(LS_HWID_SEED_KEY)
    if (existing) return existing
    const seed = window.crypto.randomUUID()
    window.localStorage.setItem(LS_HWID_SEED_KEY, seed)
    return seed
  } catch {
    // localStorage indisponível: gera um HWID novo a cada carga da página
    // (só acontece no fallback de navegador, nunca dentro do app Tauri).
    return window.crypto.randomUUID()
  }
}

async function browserHwid() {
  const seed = readOrCreateHwidSeed()
  const hex = (await sha256Hex(seed)).slice(0, 12).toUpperCase()
  return `NS-${hex.slice(0, 4)}-${hex.slice(4, 8)}-${hex.slice(8, 12)}`
}

async function browserVerifyLicenseKey(chave, hwid) {
  const key = await window.crypto.subtle.importKey(
    'spki',
    base64ToBytes(PUBLIC_KEY_SPKI_BASE64),
    { name: 'RSASSA-PKCS1-v1_5', hash: 'SHA-256' },
    false,
    ['verify'],
  )
  const signature = base64ToBytes(chave.trim())
  const message = new TextEncoder().encode(hwid.trim())
  return window.crypto.subtle.verify('RSASSA-PKCS1-v1_5', key, signature, message)
}

/** HWID desta máquina, formatado 'NS-XXXX-XXXX-XXXX' — ver TelaAtivacaoLicenca.jsx. */
export async function obterHwidMaquina() {
  if (isTauriRuntime()) {
    const { invoke } = await import('@tauri-apps/api/core')
    return invoke('obter_hwid_maquina')
  }
  return browserHwid()
}

/**
 * Ativa a licença a partir da Chave de Licença Serial colada pelo operador
 * (campo único da tela de ativação — ver TelaAtivacaoLicenca.jsx). O HWID
 * desta máquina nunca é decidido pelo lado JS: dentro do Tauri, o comando
 * Rust `ativar_serial` lê o HWID internamente e valida a assinatura RSA do
 * serial contra ele; no fallback de navegador (`npm run dev` sem o shell
 * Tauri), o HWID vem de `browserHwid()` e a verificação roda via
 * `SubtleCrypto`. Em ambos os casos, se válido, persiste a linha "oficial"
 * em `licenca_local` (status 'ATIVADO', via `licencaLocalRepo`). Lança um
 * erro com mensagem amigável quando a chave é inválida.
 */
export async function ativarSerial(serial) {
  const serialLimpo = serial.trim()
  if (!serialLimpo) throw new Error('Informe a Chave de Licença Serial.')

  let hwid
  if (isTauriRuntime()) {
    const { invoke } = await import('@tauri-apps/api/core')
    try {
      hwid = await invoke('ativar_serial', { serial: serialLimpo })
    } catch (err) {
      throw new Error(typeof err === 'string' ? err : err?.message || INVALID_KEY_MESSAGE)
    }
  } else {
    hwid = await browserHwid()
    // Chave Serial Padrão (v1.0.5): comparação normalizada, sem depender do
    // HWID — ver `isMasterKey`. Mantém o fluxo legado de assinatura RSA por
    // HWID como fallback, para chaves antigas geradas antes desta versão.
    let valid = isMasterKey(serialLimpo)
    if (!valid) {
      try {
        valid = await browserVerifyLicenseKey(serialLimpo, hwid)
      } catch {
        valid = false
      }
    }
    if (!valid) throw new Error(INVALID_KEY_MESSAGE)
  }

  await licencaLocalRepo.activate({ chaveRsa: serialLimpo, hwidMaquina: hwid })
}

/** `true` quando a licença local já está ativa — usada no boot (ver App.jsx). */
export async function verificarStatusLicenca() {
  if (isTauriRuntime()) {
    const { invoke } = await import('@tauri-apps/api/core')
    try {
      return await invoke('verificar_status_licenca')
    } catch {
      return false
    }
  }
  return licencaLocalRepo.isActive()
}

File: src-tauri/src/licensing.rs
// Etapa 5 — Módulo de Licenciamento (HWID + RSA Offline).
//
// Fluxo: `obter_hwid_maquina` deriva um identificador único e estável desta
// máquina; o cliente envia esse HWID ao suporte da NS Sistemas, que assina o
// HWID com a chave privada RSA (offline, fora deste app — ver
// `src/bin/gerar_licenca.rs`) e devolve a "Chave de Licença Serial". O app
// valida essa assinatura offline com a chave pública embutida no binário
// (`ativar_serial`) — nunca precisa de internet/servidor de licenças.
// `verificar_status_licenca` faz a checagem rápida no boot.
//
// O resultado da ativação é persistido em dois lugares, de propósito:
//   - `licenca.json` no diretório de configuração do app (gravado por este
//     módulo, em Rust puro) — é a fonte rápida e independente de SQLite
//     consultada no boot, antes de qualquer migration/conexão do
//     `tauri-plugin-sql` ter rodado.
//   - a tabela `licenca_local` (gravada pelo lado JS, via `licencaLocalRepo`
//     em `src/services/db.js`, o mesmo padrão usado por todo o resto do
//     schema) — mantém a licença auditável junto do restante dos dados da
//     loja e do backup/restauração do banco.
use rsa::pkcs8::DecodePublicKey;
use rsa::{Pkcs1v15Sign, RsaPublicKey};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::fs;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};
use tauri::Manager;
use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};

// Chave pública RSA-2048 da NS Sistemas, usada só para VERIFICAR (nunca
// assinar) Chaves de Ativação offline. A chave privada correspondente nunca
// entra neste repositório — fica só com o suporte (ver
// `src-tauri/licensing/private_key.pem`, ignorado no .gitignore, e
// `src/bin/gerar_licenca.rs`).
const LICENSE_PUBLIC_KEY_PEM: &str = include_str!("../licensing/public_key.pem");

const LICENSE_FILE_NAME: &str = "licenca.json";

#[derive(Serialize, Deserialize, Clone)]
struct LicenseRecord {
    status: String,
    hwid_maquina: String,
    chave_rsa: String,
    data_ativacao: String,
}

fn license_file_path(app: &tauri::AppHandle) -> Result<PathBuf, String> {
    let dir = app
        .path()
        .app_config_dir()
        .map_err(|error| error.to_string())?;
    Ok(dir.join(LICENSE_FILE_NAME))
}

fn read_license_record(app: &tauri::AppHandle) -> Option<LicenseRecord> {
    let path = license_file_path(app).ok()?;
    let bytes = fs::read(path).ok()?;
    serde_json::from_slice(&bytes).ok()
}

fn write_license_record(app: &tauri::AppHandle, record: &LicenseRecord) -> Result<(), String> {
    let path = license_file_path(app)?;
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|error| error.to_string())?;
    }
    let bytes = serde_json::to_vec_pretty(record).map_err(|error| error.to_string())?;
    fs::write(path, bytes).map_err(|error| error.to_string())
}

fn now_iso_seconds() -> String {
    let secs = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    // Timestamp simples (segundos desde a época Unix) — só para referência
    // interna deste marcador local; a data "oficial" de ativação exibida ao
    // usuário vem de `licenca_local.data_ativacao` (ISO 8601 completo,
    // gravado pelo lado JS em `licencaLocalRepo.activate`).
    secs.to_string()
}

/// Identificador único e estável desta máquina, formatado como
/// 'NS-XXXX-XXXX-XXXX'. No Windows deriva do MachineGuid gravado pelo
/// próprio Windows (nunca muda entre boots, só numa reinstalação do
/// sistema); nunca o valor bruto — sempre um hash SHA-256 truncado dele,
/// para não expor um identificador de sistema sensível em texto puro.
#[tauri::command]
pub fn obter_hwid_maquina(app: tauri::AppHandle) -> Result<String, String> {
    let seed = machine_seed(&app)?;
    Ok(format_hwid(&seed))
}

#[cfg(target_os = "windows")]
fn machine_seed(_app: &tauri::AppHandle) -> Result<String, String> {
    use winreg::enums::HKEY_LOCAL_MACHINE;
    use winreg::RegKey;

    let hklm = RegKey::predef(HKEY_LOCAL_MACHINE);
    let key = hklm
        .open_subkey("SOFTWARE\\Microsoft\\Cryptography")
        .map_err(|error| format!("Não foi possível ler o identificador do Windows: {error}"))?;
    let guid: String = key
        .get_value("MachineGuid")
        .map_err(|error| format!("Não foi possível ler o identificador do Windows: {error}"))?;
    Ok(guid)
}

// Fora do Windows (ambiente de desenvolvimento/testes, já que o produto é
// distribuído só para Windows): persiste um identificador aleatório estável
// no diretório de configuração do app, já que não existe MachineGuid.
#[cfg(not(target_os = "windows"))]
fn machine_seed(app: &tauri::AppHandle) -> Result<String, String> {
    use rand::RngCore;

    let dir = app
        .path()
        .app_config_dir()
        .map_err(|error| error.to_string())?;
    fs::create_dir_all(&dir).map_err(|error| error.to_string())?;
    let path = dir.join("machine_seed.txt");
    if let Ok(existing) = fs::read_to_string(&path) {
        let trimmed = existing.trim();
        if !trimmed.is_empty() {
            return Ok(trimmed.to_string());
        }
    }
    let mut bytes = [0u8; 16];
    rand::thread_rng().fill_bytes(&mut bytes);
    let seed: String = bytes.iter().map(|b| format!("{b:02x}")).collect();
    fs::write(&path, &seed).map_err(|error| error.to_string())?;
    Ok(seed)
}

fn format_hwid(seed: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(seed.as_bytes());
    let digest = hasher.finalize();
    let hex: String = digest.iter().take(6).map(|b| format!("{b:02X}")).collect();
    format!("NS-{}-{}-{}", &hex[0..4], &hex[4..8], &hex[8..12])
}

fn public_key() -> Result<RsaPublicKey, String> {
    RsaPublicKey::from_public_key_pem(LICENSE_PUBLIC_KEY_PEM.trim())
        .map_err(|error| format!("Chave pública de licenciamento inválida: {error}"))
}

fn verify_license_key(chave: &str, hwid: &str) -> Result<(), String> {
    let signature = BASE64
        .decode(chave.trim())
        .map_err(|_| "Chave de ativação inválida (formato incorreto).".to_string())?;

    let mut hasher = Sha256::new();
    hasher.update(hwid.trim().as_bytes());
    let hashed = hasher.finalize();

    public_key()?
        .verify(Pkcs1v15Sign::new::<Sha256>(), &hashed, &signature)
        .map_err(|_| "Chave de ativação não corresponde a este computador.".to_string())
}

/// Ativação por Chave de Licença Serial (tela de ativação): recebe só o
/// serial colado pelo operador, deriva o HWID desta máquina internamente
/// (nunca confia em um HWID vindo do lado JS) e valida a assinatura RSA do
/// serial contra esse HWID, 100% offline. Se válida, persiste o marcador
/// local de licença ativa (ver `LicenseRecord`) usado por
/// `verificar_status_licenca` no próximo boot, e devolve o HWID usado para
/// que o lado JS grave a linha "oficial" em `licenca_local` (SQLite) via
/// `licencaLocalRepo.activate` — ver `TelaAtivacaoLicenca.jsx`.
#[tauri::command]
pub fn ativar_serial(app: tauri::AppHandle, serial: String) -> Result<String, String> {
    let seed = machine_seed(&app)?;
    let hwid = format_hwid(&seed);

    verify_license_key(&serial, &hwid)?;

    write_license_record(
        &app,
        &LicenseRecord {
            status: "ATIVADO".to_string(),
            hwid_maquina: hwid.clone(),
            chave_rsa: serial,
            data_ativacao: now_iso_seconds(),
        },
    )?;

    Ok(hwid)
}

/// Checagem rápida no boot — lê só o marcador local (arquivo), sem depender
/// do SQLite/`tauri-plugin-sql` já estar conectado.
#[tauri::command]
pub fn verificar_status_licenca(app: tauri::AppHandle) -> Result<bool, String> {
    Ok(read_license_record(&app)
        .map(|record| record.status == "ATIVADO")
        .unwrap_or(false))
}
